### PR TITLE
[SYCL][Clang] Add support for device image compression

### PR DESCRIPTION
**Pull request by uditagarwal97**
This PR adds support for device image compression for the old offloading model. I'll make another follow-up PR to extend support for the new offload model.

**Design summary**:
```
ZSTD (compression algo)   ----> LLVMSupport (Interface)  ------> clang-offload-wrapper (For compression)
 |
 ----------------------------------------------------- --------> SYCL RT (For decompression)
```
This PR introduces ZSTD (https://github.com/facebook/zstd) as a 3rd party dependency of DPCPP. **Similar to upstream LLVM, we expect user to have `zstd-dev` package installed on their machine** - we won't be installing zstd from sources.

**How to use**
To compress device images, add `--offload-compress` CLI option to your clang invocation. Note that we compress device images only if the size of device images exceeds a threshold, which is 512 bytes by default. Moreover, by default, we use ZSTD level 10 for compression. ZSTD compression levels provides a tradeoff between (de)compression time and compression ratio, and the compression level can be changed using `--offload-compression-level=<int>` CLI option.

### CODE REVIEW

**Comment by uditagarwal97** [comment]
Some initial performance stats:

zstd_spirv_smol-v_dataset

Dataset: https://github.com/aras-p/smol-v/tree/master/tests/spirv-dumps
Dataset size: 275 SPIR-V files

**Comment by jbrodman** [comment]
What happens with the PTX and AMDGPU targets? Are they covered by the "native" binary image format? Do we need additional formats?

**Comment by jbrodman** [comment]
Also guessing this feature may not make sense when combined with the native cpu device, but need to think more about that.

**Comment by uditagarwal97** [comment]
> What happens with the PTX and AMDGPU targets? Are they covered by the "native" binary image format? Do we need additional formats?
I think they are covered by the "none" binary image format. This is because clang driver (in SYCL offload mode) never specifies the image format in call to `clang-offload-wrapper`. So, by default, the BinaryImageFormat is "none" and it is upto the SYCL runtime to determine the format (https://github.com/intel/llvm/blob/sycl/sycl/source/detail/device_binary_image.cpp#L170).

I tested my changes with PTX, and they seem to work fine, so, we'd likely not require additional formats.

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp [hunk 1]
```diff
@@ -139,15 +142,37 @@ static cl::list<std::string> Inputs(cl::Positional, cl::OneOrMore,
                                     cl::desc("<input files>"),
                                     cl::cat(ClangOffloadWrapperCategory));
 
+// CLI options for device image compression.
+// TODO: Turn off this option by default.
+static cl::opt<bool>
+    SYCLCompressDevImg("sycl-compress-dev-imgs", cl::init(true), cl::Optional,
```
Comment (sarnex) [inline clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp:148]: making sycl-specific changes here is not ideal IMO.

i wonder, could we instead do the compression as part of `sycl-post-link` and track the compression/compression type as an image property. 

maybe you already considered this and there's a reason to not do that, let me know :)

**Comment by asudarsa** [comment]
Is there some value in submitting this upstream and make it generic? Thanks

File: sycl-compress/include/sycl-compress/sycl-compress.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+#pragma once
+
+#include <memory>
+
+namespace sycl_compress {
```
Comment (bader) [inline sycl-compress/include/sycl-compress/sycl-compress.h:10]: sycl-compress library seems to be not related to SYCL at all. It also duplicated already existing library LLVM Support. We should avoid logic duplication and re-use existing library.

For the inspiration, you can take a look at offload code compression for HIP compiler - https://github.com/intel/llvm/commit/124d0b787b5d1ff4aa06bbd61a98d3fc344d0cc6.

File: sycl-compress/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+if (WIN32)
+  set(ZSTD_LIBRARY_NAME zstd_static.lib)
+else()
+  set(ZSTD_LIBRARY_NAME libzstd.a)
+endif()
+
+include(ExternalProject)
+ExternalProject_Add(zstd
+  DEPENDS
+  URL https://github.com/facebook/zstd/releases/download/v1.5.6/zstd-1.5.6.tar.gz
```
Comment (bader) [inline sycl-compress/CMakeLists.txt:27]: @uditagarwal97, please, check if the license of zstd code allows using it with our compiler.
@stdale-intel should be able to assist if needed.

File: sycl-compress/include/sycl-compress/sycl-compress.h (hunk 2 above)
Comment (uditagarwal97) [reply sycl-compress/include/sycl-compress/sycl-compress.h]: Fixed. I now link LLVMSupport library with SYCL RT and `clang-offload-wrapper` to do (de)compression.

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp (hunk 1 above)
Comment (uditagarwal97) [reply clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp]: I've renamed new CLI options added to `clang-offload-wrapper` to be sycl-agnostic: similar to the what's done in upstream `clang-offload-bundler` (https://github.com/intel/llvm/blob/sycl/clang/tools/clang-offload-bundler/ClangOffloadBundler.cpp#L180).
Also, I now link LLVMSupport library with `clang-offload-wrapper` to do the compression.

File: sycl-compress/include/sycl-compress/sycl-compress.h (hunk 2 above)
Comment (AlexeySachkov) [reply sycl-compress/include/sycl-compress/sycl-compress.h]: I wouldn't outline this code into a separate subproject, but I would maybe preserve it as a thin wrapper around `zstd` even if it is a copy of files from `LLVMSupport`.

The reason for that is one less dependency which may cause problems. You can already see that we build LLVM differently from SYCL RT on Windows and can't link the two together.

We have online compiler extension which dynamically loads `ocloc` that is linked to LLVM. Well, even our OpenCL and L0 backends are linked to LLVM. I'm not saying that it will necessarily lead to problems, but it theoretically can. An example I've found is https://stackoverflow.com/questions/29283116/is-libgcc-s-so-linked-both-statically-and-dynamically-in-the-same-process-ok

File: llvm/include/llvm/Support/Compression.h [hunk 4]
```diff
@@ -71,6 +71,8 @@ Error decompress(ArrayRef<uint8_t> Input, uint8_t *Output,
 Error decompress(ArrayRef<uint8_t> Input, SmallVectorImpl<uint8_t> &Output,
                  size_t UncompressedSize);
 
+// Get the size of the decompressed data.
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply llvm/include/llvm/Support/Compression.h:74]: PR for upstreaming: https://github.com/llvm/llvm-project/pull/107020

File: clang/include/clang/Basic/DiagnosticDriverKinds.td [hunk 5]
```diff
@@ -895,4 +895,8 @@ def err_drv_triple_version_invalid : Error<
 
 def warn_missing_include_dirs : Warning<
   "no such include directory: '%0'">, InGroup<MissingIncludeDirs>, DefaultIgnore;
+
+def warn_compress_opt_ignored : Warning<
+  "CLI option %0 ignored as it can not be used without -offload-compress">,
```
Comment (premanandrao) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td:900]:
```suggestion
  "CLI option '%0' ignored as it can not be used without '-offload-compress'">,
```

**Review by premanandrao — COMMENTED** [review]
Please add a Driver lit test to exercise the changes.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td (hunk 5 above)
Comment (uditagarwal97) [reply clang/include/clang/Basic/DiagnosticDriverKinds.td]: Fixed in 64878676eaff6c6c62a086b1aa41b0e2752a4b91

File: sycl/include/sycl/detail/compression.hpp [hunk 6]
```diff
@@ -0,0 +1,136 @@
+//==---------- compression.hpp --------------------------------------------===//
```
Comment (bader) [inline sycl/include/sycl/detail/compression.hpp:1]: Please, move this header to source directory.

Comment (uditagarwal97) [reply sycl/include/sycl/detail/compression.hpp]: Fixed in 312bd3889ebc2487693e9e9b256ade01330e20b0

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 7]
```diff
@@ -10087,6 +10087,27 @@ void OffloadWrapper::ConstructJob(Compilation &C, const JobAction &JA,
...
+      // -offload-compression-level=<>
+      if (Arg *A = C.getInputArgs().getLastArg(
+              options::OPT_offload_compression_level_EQ)) {
+
+        if (isImgCompress)
+          WrapperArgs.push_back(C.getArgs().MakeArgString(
+              Twine("-offload-compression-level=") + A->getValue()));
+      }
+    }
+
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:10110]:
```suggestion

    // Validate and propogate CLI options related to device image compression.
    // -offload-compress
    if (C.getInputArgs().getLastArg(options::OPT_offload_compress)) {
      WrapperArgs.push_back(
            C.getArgs().MakeArgString(Twine("-offload-compress")));
      // -offload-compression-level=<>
      if (Arg *A = C.getInputArgs().getLastArg(
              options::OPT_offload_compression_level_EQ))
        WrapperArgs.push_back(C.getArgs().MakeArgString(
            Twine("-offload-compression-level=") + A->getValue()));
      }
    }

```

**Comment by bader** [comment]
> **Similar to upstream LLVM, we expect user to have `zstd-dev` package installed on their machine** - we won't be installing zstd from sources.

At the same time, we should fail hard as CI does at the moment. If the library is not installed, we should just disable the compression.

File: sycl/source/CMakeLists.txt [hunk 8]
```diff
@@ -69,6 +69,15 @@ function(add_sycl_rt_library LIB_NAME LIB_OBJ_NAME)
     target_link_libraries(${LIB_NAME} PRIVATE ${ARG_XPTI_LIB})
   endif()
 
+  # TODO: Remove debug print. Need this to figure out why zstd is not
+  # being found on Windows CI machines.
+  set(CMAKE_FIND_DEBUG_MODE 1)
+
+  # Need zstd for device image compression.
+  find_package(zstd REQUIRED)
```
Comment (bader) [inline sycl/source/CMakeLists.txt:77]: The missing zstd library case should be handled gracefully.

File: sycl/test-e2e/Compression/compression.cpp [hunk 9]
```diff
@@ -0,0 +1,24 @@
+// End-to-End test for testing device image compression.
```
Comment (AlexeySachkov) [inline sycl/test-e2e/Compression/compression.cpp:1]: I would also add a test case where you have multiple translation units defining different kernels where some of those translation units are built with compression enabled and some of them were not.

NOTE: considering that compression happens at link time, you need to read "translation unit" in the example above as "fully linked library". I'm not sure if we invoke `clang-linker-wrapper` when we build static libraries, but we for sure do so for shared libraries.

**Review by AlexeySachkov — APPROVED** [review]
Compiler part looks good for me overall, just one real comment.
Runtime part could be simplified though, I think.

**Review by bso-intel — CHANGES_REQUESTED** [review]

File: sycl/source/detail/compression.hpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    // Pass ownership of the buffer to the caller.
+    return dstBuffer;
+  }
+
+  // Data fields
+private:
+  int m_lastError;
+  // ZSTD context. Reusing ZSTD context speeds up subsequent (de)compression.
+  // Storing as void* to avoid including ZSTD headers in this file.
+  void *m_ZSTD_compression_ctx;
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply sycl/source/detail/compression.hpp]: True. It was leftover code from when I had this code as a separate sub-project. Fixed now.

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp [hunk 11]
```diff
@@ -1083,10 +1108,55 @@ class BinaryWrapper {
...
+              Img.Tgt);
+        } else {
+
+          // Compress the image using zstd.
+          SmallVector<uint8_t, 512> CompressedBuffer;
+          llvm::compression::zstd::compress(
+              ArrayRef<unsigned char>(
+                  (const unsigned char *)(Bin->getBufferStart()),
+                  Bin->getBufferSize()),
+              CompressedBuffer, OffloadCompressLevel);
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp:1139]: `llvm::compression::zstd::compress` itself do error checking and throw if ran out of space or unsuccessful compression: https://llvm.org/docs/doxygen/Compression_8cpp_source.html.

If compression fails, I think it's better to throw and exit, then using the uncompressed image. This behaviour aligns with compression in `clang-offload-bundler` (what we have upstream).

[the remark this answers is not in the document]
Comment (bso-intel) [reply clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp:1139]: What kind of message is provided to the user if compression fails?
We usually surround a call to a 3rd party library call with try-catch block so we can gracefully exit with some explanation about the error.

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp [hunk 12]
```diff
@@ -1083,10 +1108,55 @@ class BinaryWrapper {
...
-            Twine(OffloadKindTag) + Twine(ImgId) + Twine(".data"), Kind,
-            Img.Tgt);
+
+        // If '--offload-compress' option is specified and zstd is not available
+        // then warn the user that the image will not be compressed.
+        if (OffloadCompressDevImgs && !llvm::compression::zstd::isAvailable()) {
+          WithColor::warning(errs(), ToolName)
+              << "'--offload-compress' option is specified but zstd is not "
+                 "available. The device image will not be compressed.\n";
+        }
```
Comment (bso-intel) [inline clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp:1118]: Also, it would be great if we are consistent about handling error conditions.
Here we gracefully give out a warning message and use uncompressed image.
But, if compression fails below, we let SYCL crash.

File: sycl/CMakeLists.txt [hunk 13]
```diff
@@ -329,6 +329,14 @@ if (LLVM_ENABLE_ASSERTIONS AND NOT SYCL_DISABLE_STL_ASSERTIONS AND NOT WIN32)
   endif()
 endif()
 
+# Need zstd for device image compression.
```
Comment (bader) [inline sycl/CMakeLists.txt:332]: Why don't use `LLVM_ENABLE_ZSTD` configured via `configure.py`?

Don't you think we should support all `LLVM_ENABLE_ZSTD` values?

> Used to decide if LLVM tools should support compression/decompression with zstd. Allowed values are OFF, ON (default, enable if zstd is found), and FORCE_ON (error if zstd is not found).

Comment (uditagarwal97) [reply sycl/CMakeLists.txt]: > Why don't use LLVM_ENABLE_ZSTD configured via configure.py?

Yup, I could use LLVM_ENABLE_ZSTD. Fixed in https://github.com/intel/llvm/pull/15124/commits/2eb0c25680be2597bb0e5a0d00ec4e1f0f3da5e0

> Don't you think we should support all LLVM_ENABLE_ZSTD values?
> 
> Used to decide if LLVM tools should support compression/decompression with zstd. Allowed values are OFF, ON (default, enable if zstd is found), and FORCE_ON (error if zstd is not found).

My understanding is that during configuration, the user assigns ON/OFF/FORCE_ON value to LLVM_ENABLE_ZSTD, but, after `find_package(zstd)` LLVM_ENABLE_ZSTD gets reassigned with TRUE/FALSE values (here: https://github.com/intel/llvm/blob/sycl/llvm/cmake/config-ix.cmake#L163). That's why, we can directly use `if (LLVM_ENABLE_ZSTD)` in CMakeFiles instead of `if (LLVM_ENABLE_ZSTD STREQUAL "FORCE_ON")` - LLVM also does the same (For example: https://github.com/intel/llvm/blob/sycl/llvm/lib/Support/Compression.cpp#L23, https://github.com/intel/llvm/blob/sycl/llvm/lib/Support/CMakeLists.txt#L28)

Comment (bader) [reply sycl/CMakeLists.txt]: Right. That's what I meant. Previous implementation implemented only "ON" value. New implementation also supports OFF and FORCE_ON values. 👍
I thought that you will just create a SYCL alias for use in the SYCL project:
```cmake
set(SYCL_ZSTD_AVAILABLE ${LLVM_ENABLE_ZSTD})
```

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp (hunk 12 above)
Comment (uditagarwal97) [reply clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp:1118]: You're right. Looking at `clang-offload-bundler`, it throws an error when ZSTD and ZLIB are not available but the user still pass `--offload-compress` CLI option to clang (https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/OffloadBundler.cpp#L1519). We should do the same with `clang-offload-wrapper`.

Comment (uditagarwal97) [reply clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp]: Fixed in https://github.com/intel/llvm/pull/15124/commits/1d8181335fb188aa4ae0ad39b3826a4162b200d2

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp:1140]: Here's LLVM's ZSTD compression function: 
```
void zstd::compress(ArrayRef<uint8_t> Input,
                    SmallVectorImpl<uint8_t> &CompressedBuffer, int Level,
                    bool EnableLdm) {
  ZSTD_CCtx *Cctx = ZSTD_createCCtx();
  if (!Cctx)
    report_bad_alloc_error("Failed to create ZSTD_CCtx");

  if (ZSTD_isError(ZSTD_CCtx_setParameter(
          Cctx, ZSTD_c_enableLongDistanceMatching, EnableLdm ? 1 : 0))) {
    ZSTD_freeCCtx(Cctx);
    report_bad_alloc_error("Failed to set ZSTD_c_enableLongDistanceMatching");
  }

  if (ZSTD_isError(
          ZSTD_CCtx_setParameter(Cctx, ZSTD_c_compressionLevel, Level))) {
    ZSTD_freeCCtx(Cctx);
    report_bad_alloc_error("Failed to set ZSTD_c_compressionLevel");
  }

  unsigned long CompressedBufferSize = ZSTD_compressBound(Input.size());
  CompressedBuffer.resize_for_overwrite(CompressedBufferSize);

  size_t const CompressedSize =
      ZSTD_compress2(Cctx, CompressedBuffer.data(), CompressedBufferSize,
                     Input.data(), Input.size());

  ZSTD_freeCCtx(Cctx);

  if (ZSTD_isError(CompressedSize))
    report_bad_alloc_error("Compression failed");

  __msan_unpoison(CompressedBuffer.data(), CompressedSize);
  if (CompressedSize < CompressedBuffer.size())
    CompressedBuffer.truncate(CompressedSize);
}
```
LLVM throws the appropriate error message, depending on the type of error.

> We usually surround a call to a 3rd party library call with try-catch block so we can gracefully exit with some explanation about the error.

While I totally agree with it but since we are using upstream code for image compression, I don't think we should deviate from the upstream in terms of error handling.

**Comment by uditagarwal97** [comment]
Ping @premanandrao @mdtoguchi @bso-intel

File: sycl/test-e2e/lit.cfg.py [hunk 14]
```diff
@@ -308,6 +308,33 @@ def open_check_file(file_name):
 if sp[0] == 0:
     config.available_features.add("preview-breaking-changes-supported")
 
+# Check if clang is built with ZSTD and compression support.
```
Comment (AlexeySachkov) [inline sycl/test-e2e/lit.cfg.py:311]: There is a way simpler way. Use [`lit.site.cfg.py.in`](https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/lit.site.cfg.py.in) to propagate a value of CMake variable into this python script.

I would also explore how LLVM propagates that. There are tests in LLVM which require `zstd` feature, so I wonder if we can call some LIT helper to get this feature automatically propagated into LIT for us

File: sycl/test-e2e/Compression/Inputs/single_kernel.cpp [hunk 15]
```diff
@@ -0,0 +1,17 @@
+#include <sycl/detail/core.hpp>
+
+int main() {
+
+  sycl::queue q0;
+  int val = -1;
+  {
+    sycl::buffer<int, 1> buffer1(&val, sycl::range(1));
+
+    q0.submit([&](sycl::handler &cgh) {
```
Comment (cperkinsintel) [inline sycl/test-e2e/Compression/Inputs/single_kernel.cpp:10]: The description on this PR says that there is a threshold of 512 bytes, below which we don't compress.

Is this tiny kernel above or below that threshold?  And shouldn't we have a test for the other side of that threshold as well?

File: sycl/source/detail/compression.hpp [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+  // Get the singleton instance of the ZSTDCompressor class.
+  static ZSTDCompressor &GetSingletonInstance() {
+    static ZSTDCompressor instance;
+    return instance;
+  }
+
+  // Public APIs
+public:
+  // Blob (de)compression do not assume format/structure of the input buffer.
+  static std::unique_ptr<char> CompressBlob(const char *src, size_t srcSize,
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply sycl/source/detail/compression.hpp:43]: While we don't use `CompressBlob` for device image compression, it does help in writing unit tests and could be helpful, if in the future, we decide to compress images in kernel program cache (just thinking out loud).

File: sycl/test-e2e/Compression/Inputs/single_kernel.cpp (hunk 15 above)
Comment (uditagarwal97) [reply sycl/test-e2e/Compression/Inputs/single_kernel.cpp:10]: > Is this tiny kernel above or below that threshold? 

It is above the threshold maybe because I use `-O0` to compile :P

>  And shouldn't we have a test for the other side of that threshold as well?

Yes, I have the following test to ensure that there's no compression when the size < threshold:
[clang/test/Driver/clang-offload-wrapper-zstd.c](https://github.com/intel/llvm/pull/15124/files#diff-a8f20e0fdb0b70c22204d8c7fb22aeb6d3d9b65258e5c6a8e9223db699e3cd95)

File: sycl/test-e2e/Compression/compression_seperate_compile.cpp [hunk 17]
```diff
@@ -0,0 +1,70 @@
+// End-to-End test for testing device image compression when we
+// seperatly compile and link device images.
+
+// REQUIRES: zstd, opencl-aot, cpu, linux
+
+//////////////////////  Compile device images
+// RUN: %clangxx -fsycl -fsycl-targets=spir64_x86_64 -fsycl-host-compiler=clang++ -fsycl-host-compiler-options='-std=c++17 -Wno-attributes -Wno-deprecated-declarations -fPIC -DENABLE_KERNEL1' -DENABLE_KERNEL1 -c %s -o %t_kernel1_aot.o
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply sycl/test-e2e/Compression/compression_seperate_compile.cpp]: I wanted to have a test that mimics the compilation toolchain that PyTorch team use (as described here: https://github.com/intel/torch-xpu-ops/blob/main/cmake/BuildFlags.cmake#L63). They use `gcc` for final linkage. The problem with using `gcc` here in E2E test is that we'd have to explicitly provide path to sycl headers and library (See the older version of this test: https://github.com/intel/llvm/blob/1d8181335fb188aa4ae0ad39b3826a4162b200d2/sycl/test-e2e/Compression/compression_seperate_compile.cpp). AFAIK, we don't have LIT substitutions to get path to SYCL headers and library, and so, I ended up using clang++ as "3rd party" compiler, with which I can just use `-fsycl` to get the include directory and SYCL library.

File: sycl/test-e2e/Compression/compression_seperate_compile.cpp [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+#include <sycl/usm.hpp>
+
+#include <iostream>
+
+class kernel_init;
+void run_kernel1(int *a, queue q);
+void run_kernel2(int *a, queue q);
+int main() {
+  int retCode = 0;
+  queue q(cpu_selector_v, {property::queue::in_order()});
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply sycl/test-e2e/Compression/compression_seperate_compile.cpp]: True. Fixed.

File: sycl/test-e2e/lit.cfg.py (hunk 14 above)
Comment (uditagarwal97) [reply sycl/test-e2e/lit.cfg.py:311]: As discussed offline, passing `LLVM_ENABLE_ZSTD` from CMake to LIT won't work here because E2E tests can be built standalone, like what we do in CI.
LLVM seems to pass CMake Variables to LIT: https://github.com/llvm/llvm-project/blob/main/llvm/test/lit.site.cfg.py.in#L37

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 19]
```diff
@@ -10131,6 +10131,19 @@ void OffloadWrapper::ConstructJob(Compilation &C, const JobAction &JA,
...
+    // -offload-compress
+    if (C.getInputArgs().getLastArg(options::OPT_offload_compress)) {
+      WrapperArgs.push_back(
+          C.getArgs().MakeArgString(Twine("-offload-compress")));
+      // -offload-compression-level=<>
+      if (Arg *A = C.getInputArgs().getLastArg(
+              options::OPT_offload_compression_level_EQ))
+        WrapperArgs.push_back(C.getArgs().MakeArgString(
+            Twine("-offload-compression-level=") + A->getValue()));
+    }
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:10145]: Could you add a driver specific test that verifies the options being passed to the `clang-offload-wrapper`?

File: sycl/test-e2e/Compression/no_zstd_warning.cpp [hunk 20]
```diff
@@ -0,0 +1,5 @@
+// using --offload-compress without zstd should throw an error.
+// REQUIRES: !zstd
+// RUN: %{build} -O0 -g --offload-compress %S/Inputs/single_kernel.cpp -o %t_compress.out 2>&1 | FileCheck %s
+
+// XFAIL: *
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply sycl/test-e2e/Compression/no_zstd_warning.cpp]: Fixed in 58f9939c2dffefab063c13e2caa34b93caa9b39e

File: clang/test/Driver/sycl-offload-wrapper-compression.cpp [hunk 21]
```diff
@@ -0,0 +1,16 @@
+///
+/// Check if '--offload-compress' and '--offload-compression-level' CLI
+/// options are passed to the clang-offload-wrapper.
+///
+
+// REQUIRES: zstd
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-offload-wrapper-compression.cpp:6]: As this is a driver specific test just checking the `-###` there should be no zstd requirement for it to run.

File: sycl/test-e2e/lit.cfg.py (hunk 14 above)
Comment (AlexeySachkov) [reply sycl/test-e2e/lit.cfg.py:311]: Yeah, good point. I still wonder if there is a simpler way (i.e. some existing helper for running an executable and getting its output, but if no, then we will have to leave with what we have.

BTW, I'm sure that compiler is able to read the program from `stdin`, so you can maybe save on file operations here

File: clang/test/Driver/sycl-offload-wrapper-compression.cpp (hunk 21 above)
Comment (uditagarwal97) [reply clang/test/Driver/sycl-offload-wrapper-compression.cpp]: Fixed in https://github.com/intel/llvm/pull/15124/commits/970ad3563efe05d7c65bf0b54050de63de645be1

**Review by mdtoguchi — APPROVED** [review]
OK for driver

**Comment by uditagarwal97** [comment]
@bso-intel @intel/llvm-reviewers-runtime ping!

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (bso-intel) [reply clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp:1141]: When the crash occurs when compression failed, the error message that the SYCL end-user will  receive may not be useful.
For example, "Failed to create ZSTD_CCtx", ""Failed to set ZSTD_c_compressionLevel", etc.
It would be much better user experience if they get a message like "Device image compression failed." + e.what().

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp [hunk 22]
```diff
@@ -1083,10 +1108,57 @@ class BinaryWrapper {
...
-            Img.Tgt);
+
+        // If '--offload-compress' option is specified and zstd is not
+        // available, throw an error.
+        if (OffloadCompressDevImgs && !llvm::compression::zstd::isAvailable()) {
+          return createStringError(
+              inconvertibleErrorCode(),
+              "'--offload-compress' option is specified but zstd "
+              "is not available. The device image will not be "
+              "compressed.");
```
Comment (bso-intel) [inline clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp:1119]: This kind of error message is good since the SYCL end-user can understand what went wrong.

**Review by bso-intel — CHANGES_REQUESTED** [review]

**Review by cperkinsintel — APPROVED** [review]

**Comment by uditagarwal97** [comment]
@bso-intel 

> When the crash occurs when compression failed, the error message that the SYCL end-user will receive may not be useful.
> For example, "Failed to create ZSTD_CCtx", ""Failed to set ZSTD_c_compressionLevel", etc.
> It would be much better user experience if they get a message like "Device image compression failed." + e.what().

In https://github.com/intel/llvm/pull/15124/commits/946a738d5b7730e5af1cdf56d4c0ceaa176991da, I've wrapped zstd::compress in try/catch to throw a more meaningful error message. Note that this will only work if DPC++ is built with `LLVM_ENABLE_EH`.

**Review by bso-intel — APPROVED** [review]

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 23]
```diff
@@ -731,6 +733,14 @@ setSpecializationConstants(const std::shared_ptr<device_image_impl> &InputImpl,
   }
 }
 
+static inline void CheckAndDecompressImage(RTDeviceBinaryImage *Img) {
```
Comment (uditagarwal97) [inline sycl/source/detail/program_manager/program_manager.cpp:736]: add `[[maybe_unused]]`

**Comment by uditagarwal97** [comment]
@intel/llvm-gatekeepers The PR is ready to be merged. All the downstream infrastructure, along with intel/llvm CI machines, is ready with zstd installed.

**Comment by againull** [comment]
@uditagarwal97 Could you please take a look at post-commit failures: https://github.com/intel/llvm/actions/runs/11468699849/job/31915329901

File: buildbot/configure.py [hunk 24]
```diff
@@ -178,6 +178,8 @@ def do_configure(args):
         "-DLLVM_ENABLE_PROJECTS={}".format(llvm_enable_projects),
         "-DSYCL_BUILD_PI_HIP_PLATFORM={}".format(sycl_build_pi_hip_platform),
         "-DLLVM_BUILD_TOOLS=ON",
+        "-DLLVM_ENABLE_ZSTD=ON",
```
Comment (jsji) [inline buildbot/configure.py:181]: I think we shouldn't turn on these by default? We should just turn on them in `if args.ci_defaults:`

**Comment by uditagarwal97** [comment]
> @uditagarwal97 Could you please take a look at post-commit failures: https://github.com/intel/llvm/actions/runs/11468699849/job/31915329901
> 
> Failed Tests (2): SYCL :: Compression/compression.cpp SYCL :: Compression/compression_multiple_tu.cpp
> 
> On AMD/HIP

PR to disable the failing tests on HIP: https://github.com/intel/llvm/pull/15830

File: buildbot/configure.py (hunk 24 above)
Comment (uditagarwal97) [reply buildbot/configure.py:181]: The idea behind turning them on by default is to build the compiler with device image compression support if the user has zstd-dev package installed. If zstd is not found, there shouldn't be any build error.

Comment (jsji) [reply buildbot/configure.py:181]: there are configurate failures.

Comment (jsji) [reply buildbot/configure.py:181]: CMake Error at lib/Support/CMakeLists.txt:327 (get_property):

  get_property could not find TARGET zstd::libzstd_static.  Perhaps it has

  not yet been created.



CMake Error at lib/Support/CMakeLists.txt:330 (get_property):

  get_property could not find TARGET zstd::libzstd_static.  Perhaps it has

  not yet been created.

Comment (sarnex) [reply buildbot/configure.py:181]: yeah local build is cooked for me, we need to revert the on by zstd default part at least

Comment (uditagarwal97) [reply buildbot/configure.py:181]: PR to temporarily turn off zstd by default: https://github.com/intel/llvm/pull/15833
